clients/src/main/java/org/apache/kafka/common/record/AbstractRecords.java

```diff
@@ -52,51 +53,96 @@ public boolean hasCompatibleMagic(byte magic) {
      * Convert this message set to use the specified message format.
      */
     @Override
-    public Records toMessageFormat(byte toMagic, TimestampType upconvertTimestampType) {
-        List<LogRecord> records = Utils.toList(records().iterator());
-        if (records.isEmpty()) {
+    public Records toMagic(byte magic, TimestampType upconvertTimestampType) {
+        List<? extends LogEntry> entries = Utils.toList(entries().iterator());
+        if (entries.isEmpty()) {
             // This indicates that the message is too large, which indicates that the buffer is not large
             // enough to hold a full log entry. We just return all the bytes in the file message set.
             // Even though the message set does not have the right format version, we expect old clients
             // to raise an error to the user after reading the message size and seeing that there
             // are not enough available bytes in the response to read the full message.
             return this;
         } else {
-            // We use the first message to determine the compression type for the resulting message set.
-            // This could result in message sets which are either larger or smaller than the original size.
-            // For example, it could end up larger if most messages were previously compressed, but
-            // it just so happens that the first one is not. There is also some risk that this can
-            // cause some timestamp information to be lost (e.g. if the timestamp type was changed) since
-            // we are essentially merging multiple message sets. However, currently this method is only
-            // used for down-conversion, so we've ignored the problem.
-            LogEntry firstEntry = entries().iterator().next();
-
-            // FIXME: Using the timestamp and compression type from the first message is almost certainly wrong
-            //        We also need to take into account pid information
-            TimestampType timestampType = firstEntry.timestampType();
-            if (timestampType == TimestampType.NO_TIMESTAMP_TYPE)
-                timestampType = upconvertTimestampType;
-
-            return convert(toMagic, timestampType, firstEntry.compressionType(), firstEntry.timestamp(), records);
+            List<LogEntryAndRecords> logEntryAndRecordsList = new ArrayList<>(entries.size());
+            int totalSizeEstimate = 0;
+            boolean upconvertNonCompressed = magic >= LogEntry.MAGIC_VALUE_V2;
+
+            for (LogEntry entry : entries) {
+                List<LogRecord> logRecords = Utils.toList(entry.iterator());
+                final long baseOffset;
+                if (entry.magic() >= LogEntry.MAGIC_VALUE_V2)
+                    baseOffset = entry.baseOffset();
+                else
+                    baseOffset = logRecords.get(0).offset();
+                totalSizeEstimate += sizeEstimateInBytes(magic, baseOffset, entry.compressionType(), logRecords);
```

Comment:
I'm not too sure. When using compression, the estimate is already very fuzzy. But if the cost of the extra diligence is not too high, it seems worthwhile getting an exact estimate for the non-compressed path.